File: app/src/main/java/com/skyd/imomoe/util/dlna/dmc/control/ControlImpl.java
package com.skyd.imomoe.util.dlna.dmc.control;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.skyd.imomoe.util.dlna.dmc.ICast;
import com.skyd.imomoe.util.dlna.dmc.Utils;

import org.fourthline.cling.controlpoint.ControlPoint;
import org.fourthline.cling.model.meta.Device;

import java.util.Map;

public class ControlImpl implements ICastInterface.IControl {

    private final IServiceFactory mServiceFactory;
    private final Device<?, ?, ?> mDevice;
    private final Map<String, IServiceAction.IServiceActionCallback<?>> mCallbackMap;

    public ControlImpl(@NonNull ControlPoint controlPoint, @NonNull Device<?, ?, ?> device, Map<String, IServiceAction.IServiceActionCallback<?>> map) {
        mDevice = device;
        mCallbackMap = map;
        mServiceFactory = new IServiceFactory.ServiceFactoryImpl(controlPoint, device);
    }

    @Override
    public void cast(Device<?, ?, ?> device, ICast object) {
        mServiceFactory.getAvService().cast(new ICastInterface.CastEventListener() {
            @Override
            public void onSuccess(String result) {
                IServiceAction.IServiceActionCallback<Object> listener = getCallback(IServiceAction.ServiceAction.CAST);
                if (listener != null) listener.onSuccess(result);
            }

            @Override
            public void onFailed(String errMsg) {
                IServiceAction.IServiceActionCallback<Object> listener = getCallback(IServiceAction.ServiceAction.CAST);
                if (listener != null) listener.onFailed(errMsg);
            }
        }, object.getUri(), Utils.getMetadata(object));
    }

    @Override
    public boolean isCasting(Device<?, ?, ?> device) {
        return mDevice != null && mDevice.equals(device);
    }

    @Override
    public void stop() {
        mServiceFactory.getAvService().stop(getCallback(IServiceAction.ServiceAction.STOP));
    }

    @Override
    public void play() {
        mServiceFactory.getAvService().play(getCallback(IServiceAction.ServiceAction.PLAY));
    }

    @Override
    public void pause() {
        mServiceFactory.getAvService().pause(getCallback(IServiceAction.ServiceAction.PAUSE));
    }

    @Override
    public void seekTo(long position) {
        mServiceFactory.getAvService().seek(getCallback(IServiceAction.ServiceAction.SEEK_TO), position);
    }

    @Override
    public void setVolume(int percent) {
        mServiceFactory.getRenderService().setVolume(getCallback(IServiceAction.ServiceAction.SET_VOLUME), percent);
    }

    @Override
    public void setMute(boolean mute) {
        mServiceFactory.getRenderService().setMute(getCallback(IServiceAction.ServiceAction.SET_MUTE), mute);
    }

    @Override
    public void setBrightness(int percent) {
        mServiceFactory.getRenderService().setBrightness(getCallback(IServiceAction.ServiceAction.SET_BRIGHTNESS), percent);
    }

    @SuppressWarnings("unchecked")
    @Nullable
    private <T> IServiceAction.IServiceActionCallback<T> getCallback(IServiceAction.ServiceAction action) {
        Object result = mCallbackMap.get(action.name());
        if (result == null) return null;
        return (IServiceAction.IServiceActionCallback<T>) result;
    }
}

File: app/src/main/java/com/skyd/imomoe/util/dlna/dmc/control/ControlImpl.kt
package com.skyd.imomoe.util.dlna.dmc.control

import org.fourthline.cling.controlpoint.ControlPoint
import com.skyd.imomoe.util.dlna.dmc.control.IServiceAction.IServiceActionCallback
import com.skyd.imomoe.util.dlna.dmc.control.ICastInterface.IControl
import com.skyd.imomoe.util.dlna.dmc.ICast
import com.skyd.imomoe.util.dlna.dmc.Utils
import com.skyd.imomoe.util.dlna.dmc.control.ICastInterface.CastEventListener
import com.skyd.imomoe.util.dlna.dmc.control.IServiceAction.ServiceAction
import com.skyd.imomoe.util.dlna.dmc.control.IServiceFactory.ServiceFactoryImpl
import org.fourthline.cling.model.meta.Device

class ControlImpl(
    val controlPoint: ControlPoint,
    val device: Device<*, *, *>,
    val callbackMap: Map<String, IServiceActionCallback<*>>
) : IControl {
    private val serviceFactory: IServiceFactory

    override fun cast(device: Device<*, *, *>, `object`: ICast) {
        serviceFactory.avService.cast(object : CastEventListener {
            override fun onSuccess(result: String) {
                val listener = getCallback<Any>(ServiceAction.CAST)
                listener?.onSuccess(result)
            }

            override fun onFailed(errMsg: String) {
                val listener = getCallback<Any>(ServiceAction.CAST)
                listener?.onFailed(errMsg)
            }
        }, `object`.uri, Utils.getMetadata(`object`))
    }

    override fun isCasting(device: Device<*, *, *>): Boolean {
        return this.device == device
    }

    override fun stop() {
        serviceFactory.avService.stop(getCallback(ServiceAction.STOP))
    }

    override fun play() {
        serviceFactory.avService.play(getCallback(ServiceAction.PLAY))
    }

    override fun pause() {
        serviceFactory.avService.pause(getCallback(ServiceAction.PAUSE))
    }

    override fun seekTo(position: Long) {
        serviceFactory.avService.seek(getCallback(ServiceAction.SEEK_TO), position)
    }

    override fun setVolume(percent: Int) {
        serviceFactory.renderService.setVolume(getCallback(ServiceAction.SET_VOLUME), percent)
    }

    override fun setMute(mute: Boolean) {
        serviceFactory.renderService.setMute(getCallback(ServiceAction.SET_MUTE), mute)
    }

    override fun setBrightness(percent: Int) {
        serviceFactory.renderService
            .setBrightness(getCallback(ServiceAction.SET_BRIGHTNESS), percent)
    }

    private fun <T> getCallback(action: ServiceAction): IServiceActionCallback<T>? {
        val result = callbackMap[action.name] ?: return null
        return result as IServiceActionCallback<T>
    }

    init {
        serviceFactory = ServiceFactoryImpl(controlPoint, device)
    }
}